node/lib/stitch.js

```diff
@@ -0,0 +1,243 @@
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+const ArgumentParser = require("argparse").ArgumentParser;
+const co             = require("co");
+const NodeGit        = require("nodegit");
+
+const DoWorkQueue         = require("./util/do_work_queue");
+const GitUtil             = require("./util/git_util");
+const StitchUtil          = require("./util/stitch_util");
+
+const description = `Stitch together the specified meta-repo commitish in \
+the specified repo.`;
+
+const parser = new ArgumentParser({
+    addHelp: true,
+    description: description
+});
+
+parser.addArgument(["--no-fetch"], {
+    required: false,
+    action: "storeConst",
+    constant: true,
+    defaultValue: false,
+    help: `If provided, assume commits are present and do not fetch.`,
+});
+
+parser.addArgument(["-t", "--target-branch"], {
+    required: false,
+    type: "string",
+    defaultValue: "master",
+    help: "Branch to update with committed ref; default is 'master'.",
+});
+
+parser.addArgument(["-j"], {
+    required: false,
+    type: "int",
+    help: "number of parallel operations, default 8",
+    defaultValue: 8,
+});
+
+parser.addArgument(["-c", "--commitish"], {
+    type: "string",
+    help: "meta-repo commit to stitch, default is HEAD",
+    defaultValue: "HEAD",
+    required: false,
+});
+
+parser.addArgument(["-u", "--url"], {
+    type: "string",
+    defaultValue: null,
+    help: `location of the origin repository where submodules are rooted, \
+required unless '--no-fetch' is specified`,
+    required: false,
+});
+
+parser.addArgument(["-r", "--repo"], {
+    type: "string",
+    help: "location of the repo, default is \".\"",
+    defaultValue: ".",
+});
+
+parser.addArgument(["-k", "--keep-as-submodule"], {
+    type: "string",
+    help: `submodules whose paths are matched by this regex are not stitched, \
+but are instead kept as submodules.`,
+    required: false,
+    defaultValue: null,
+});
+
+parser.addArgument(["--skip-empty"], {
+    required: false,
+    action: "storeConst",
+    constant: true,
+    defaultValue: false,
+    help: "Skip a commit if its  tree would be the same as its first parent.",
+});
+
+parser.addArgument(["--root"], {
+    type: "string",
+    help: "When provided, run a *join* operation that creates a new history \
+joining those of the submodules under the specified 'root' path.  \
+In this history those paths will be relative to 'root', i.e., they will not \
+be prefixed by it. This option implies '--skip-empty'.",
+});
+
+const stitch = co.wrap(function *(repoPath,
```

Comment:
Should we wrap some of these args into an options object?

Reply:
Yes, I think I'll do that and also move the function into `stitch_util` and give it a test driver.